### PR TITLE
Add support for signed tasks.

### PR DESCRIPTION
**Pull request by hooverdc**

### CODE REVIEW

File: django_tasks/backends/database/models.py [hunk 1]
```diff
@@ -63,7 +63,7 @@ class DBTaskResult(GenericBase[P, T], models.Model):
         max_length=max(len(value) for value in ResultStatus.values),
     )
 
-    enqueued_at = models.DateTimeField(auto_now_add=True)
```
Comment (RealOrangeOne) [inline django_tasks/backends/database/models.py:66]: Question: What's the reason for this change?

File: django_tasks/backends/base.py [hunk 2]
```diff
@@ -27,6 +27,9 @@ class BaseTaskBackend(metaclass=ABCMeta):
     supports_get_result = False
     """Can results be retrieved after the fact (from **any** thread / process)"""
 
+    supports_signed_task = False
```
Comment (RealOrangeOne) [inline django_tasks/backends/base.py:30]: Suggestion: Because signing is strictly an internal operation, I don't think this needs to be something a backend "supports".

File: django_tasks/backends/database/backend.py [hunk 3]
```diff
@@ -28,13 +30,14 @@ class DatabaseBackend(BaseTaskBackend):
...
     supports_defer = True
+    supports_signed_task = True
 
     def _task_to_db_task(
         self, task: Task[P, T], args: P.args, kwargs: P.kwargs
     ) -> "DBTaskResult":
         from .models import DBTaskResult
 
-        return DBTaskResult(
+        db_task_result: DBTaskResult = DBTaskResult(
```
Comment (RealOrangeOne) [inline django_tasks/backends/database/backend.py:40]: Nit: The explicit type shouldn't be necessary here
```suggestion
        db_task_result = DBTaskResult(
```

File: django_tasks/backends/database/backend.py [hunk 4]
```diff
@@ -43,6 +46,17 @@ def _task_to_db_task(
             backend_name=self.alias,
         )
 
+        # sign value without timestamp
+        # TimestampSigner doesn't support setting our own timestamp?
+        if settings.TASKS.get(self.alias, {}).get("SIGN_TASKS") is True:
```
Comment (RealOrangeOne) [inline django_tasks/backends/database/backend.py:51]: Issue: This is the wrong way to get the settings for the backend. Settings should be unpacked in the backend's constructor

File: django_tasks/backends/database/models.py [hunk 5]
```diff
@@ -124,6 +126,30 @@ def task_result(self) -> "TaskResult[T]":
 
         return result
 
+    @property
+    def canonical(self) -> Dict:
```
Comment (RealOrangeOne) [inline django_tasks/backends/database/models.py:130]: Suggestion: This might be something better to store on the `TaskResult` instead. That way, it's not specific to the DB implementation.

File: django_tasks/backends/database/backend.py [hunk 6]
```diff
@@ -43,6 +46,17 @@ def _task_to_db_task(
...
+            signer = Signer()
+            _, _, signature = signer.sign_object(db_task_result.canonical).partition(
```
Comment (RealOrangeOne) [inline django_tasks/backends/database/backend.py:53]: Suggestion: In much the same way that user password verification is done, I think signature creation and validation should live on the `TaskResult` - that way the implementation can be shared, and kept in 1 place.

Alternatively, perhaps as separate methods on the backend itself?

File: django_tasks/backends/database/backend.py [hunk 7]
```diff
@@ -43,6 +46,17 @@ def _task_to_db_task(
             backend_name=self.alias,
         )
 
+        # sign value without timestamp
+        # TimestampSigner doesn't support setting our own timestamp?
+        if settings.TASKS.get(self.alias, {}).get("SIGN_TASKS") is True:
+            signer = Signer()
```
Comment (RealOrangeOne) [inline django_tasks/backends/database/backend.py:52]: Issue: The signature should be salted. For compatibility, we should also specify the `algorithm`, as this could change in future Django versions, leading to signature issues.

File: django_tasks/backends/database/management/commands/db_worker.py [hunk 8]
```diff
@@ -96,6 +99,16 @@ def run_task(self, db_task_result: DBTaskResult) -> None:
             task = db_task_result.task
             task_result = db_task_result.task_result
 
+            if settings.TASKS.get(self.backend_name, {}).get("SIGN_TASKS") is True:
+                signer = Signer()
+                _, _, signature = signer.sign_object(
+                    db_task_result.canonical
+                ).partition(":")
+                if db_task_result.signature != signature:
```
Comment (RealOrangeOne) [inline django_tasks/backends/database/management/commands/db_worker.py:107]: Question: Should this be a constant-time comparison?

File: django_tasks/backends/database/models.py [hunk 9]
```diff
@@ -124,6 +126,30 @@ def task_result(self) -> "TaskResult[T]":
...
+        """
+        Get canonical form
+        """
+        return {
```
Comment (RealOrangeOne) [inline django_tasks/backends/database/models.py:134]: Suggestion: Because `TaskResult` is a dataclass, you can call `asdict` on it to get it as a dict natively, and then massage the types into something JSON-serializable.

**Review by RealOrangeOne — COMMENTED** [review]
Thanks so much for this feature! It's something we spoke of during the DEP process, but ended up working around with opt-in task functions. However, I do think it'd be a nice feature to add.

Along with my comments, it'd be great if you could add some tests for this new functionality!

**Comment by hooverdc** [comment]
Thanks, I'll get working on those.

File: django_tasks/backends/database/models.py (hunk 1 above)
Comment (hooverdc) [reply django_tasks/backends/database/models.py:66]: `enqueued_at` was getting set at insert rather than at model creation time. We would have to do an insert and an update to set the signature. Now whether or not enqueued_at should represented the time `.enqueue` was called or the time the record was inserted, I'm not sure about.

File: django_tasks/backends/database/backend.py (hunk 3 above)
Comment (hooverdc) [reply django_tasks/backends/database/backend.py:43]: mypy yelled about it, it really should be able to infer that.

File: django_tasks/backends/database/models.py (hunk 9 above)
Comment (hooverdc) [reply django_tasks/backends/database/models.py]: I moved `.canonical` inside the base task result, but for task I only wanted to include invariant values. I don't think status, started_at, or finished_at should form part of the signature.

**Comment by hooverdc** [comment]
@RealOrangeOne Added tests and made updates based on your suggestions.

**Comment by RealOrangeOne** [comment]
CI is still failing. However, on reflection, I'm not sure the added complexity is worthwhile. 

Because task functions must be explicitly allowed, there's no ability to gain remote-code execution inside Django (at least not in ways which can be controlled by `django-tasks`). Whilst it's possible to potentially modify data, if an attacker already has access to the database they can do that anyway, not to mention that they probably have access to other parts of your infrastructure too.
